ofnetSvcProxy.go

```diff
@@ -0,0 +1,662 @@
+	}
+
+	// Clean up operational state
+	proxy.oMutex.Lock()
+	defer proxy.oMutex.Unlock()
+	oper := proxy.operState
+	operEntry, found := oper[spec.IpAddress]
+	if !found {
+		log.Debugf("delService operEntry not found for %s", svcName)
+		return
+	}
+
+	// delete the watched flows from OVS
+	for _, flow := range operEntry.watchedFlows {
+		if flow != nil {
+			flow.Delete()
+		}
+	}
+
+	// delete the nat'ed flows
+	for key, flow := range operEntry.natFlows {
+		if flow != nil {
+			flow.Delete()
+			log.Infof("NAT flow %s deleted", key)
+		} else {
+			log.Infof("NAT flow %s not found", key)
+		}
+	}
+
+	// remove the operEntry
+	delete(oper, spec.IpAddress)
+}
+
+// AddSvcSpec adds or updates a service spec.
+func (proxy *ServiceProxy)AddSvcSpec(svcName string, spec *ServiceSpec) error {
+	log.Infof("AddSvcSpec %s %v", svcName, spec)
+	// Make sure spec is valid
+	for _, p := range spec.Ports {
+		if p.Protocol != "TCP" && p.Protocol != "UDP" {
+			return errors.New("Invalid protocol in spec")
+		}
+	}
+
+	services := proxy.catalogue.svcMap
+	oldSpec, found := services[svcName]
+	if found {
+		if matchSpec(&oldSpec, spec) {
+			log.Debugf("No change in spec for %s", svcName)
+			return nil
+		}
+
+		// delete existing spec and add the new one
+		proxy.DelSvcSpec(svcName, &oldSpec)
+	}
+
+	services[svcName] = *spec
+	proxy.addService(svcName)	
+	return nil
+}
+
+// DelSvcSpec deletes a service spec.
+func (proxy *ServiceProxy)DelSvcSpec(svcName string, spec *ServiceSpec) {
+	log.Infof("DelSvcSpec %s %v", svcName, spec)
+	services := proxy.catalogue.svcMap
+	_, found := services[svcName]
+	if !found {
+		log.Warnf("DelSvcSpec service %s not found", svcName)
+	} else {
+		proxy.delService(svcName)
+		delete(services, svcName)
+	}
+}
+
+// addProvider adds the given provider to operational State
+func (proxy *ServiceProxy)addProvider(svcIP, provIP string) error {
+	oper := proxy.operState
+        proxy.oMutex.Lock()
+        defer proxy.oMutex.Unlock()
+        operEntry, found := oper[svcIP]
+        if !found {
+                log.Errorf("addProvider operEntry not found for %s", svcIP)
+                return errors.New("operEntry not found")
+        }
+	operEntry.addProvHdl(provIP)
+	log.Infof("Added provider %s for serviceIP %s", provIP, svcIP)
+	return nil
+}
+
+// delProvider deletes the given provider from operational State
+func (proxy *ServiceProxy)delProvider(svcIP, provIP string) error {
+	oper := proxy.operState
+        proxy.oMutex.Lock()
+        defer proxy.oMutex.Unlock()
+        operEntry, found := oper[svcIP]
+        if !found {
+                log.Errorf("delProvider operEntry not found for %s", svcIP)
+                return errors.New("operEntry not found")
+        }
+
+	// Remove flows NAT'ed to this provider
+	for epIP, _ := range operEntry.provHdl[provIP].clientEPs {
+		for _, p := range operEntry.ports {
+			operEntry.delNATFlow(epIP, "Dst", &p)
+			operEntry.delNATFlow(epIP, "Src", &p)
+		}
+	}
```

Comment:
That will automatically happen when the client accesses the service via the flow miss path. No need for special handling.